File: src/scanner.rs
use std::str::FromStr;

use crate::{
    error::{Error, Result},
    token::{Token, TokenKind},
};

pub struct Scanner<'a> {
    start: usize,
    current: usize,
    line: usize,
    column: usize,
    source: &'a str,
}

impl Scanner<'_> {
    pub fn get_tokens(source: &str) -> Result<Vec<Token>> {
        let mut scanner = Scanner {
            start: 0,
            current: 0,
            line: 1,
            column: 0,
            source,
        };
        let mut res = vec![];
        loop {
            let token = scanner.next_token()?;
            // println!("token: {:?}", token);
            let kind = token.kind.clone();
            res.push(token);
            match kind {
                TokenKind::Eof => break,
                _ => {}
            }
        }
        Ok(res)
    }

    fn err(&mut self, msg: String) -> Error {
        Error {
            line: self.line,
            start: self.start,
            end: self.current,
            msg,
        }
    }

    fn next_token(&mut self) -> Result<Token> {
        self.skip_whitespace()?;
        self.start = self.current;
        if self.is_at_end() {
            return self.make_token(TokenKind::Eof);
        }

        let c = self.advance().unwrap();

        match c {
            'a'..='z' | 'A'..='Z' => self.identifier(),
            '0'..='9' => self.number(),
            '(' => self.make_token(TokenKind::LeftParen),
            ')' => self.make_token(TokenKind::RightParen),
            '[' => self.make_token(TokenKind::LeftBracket),
            ']' => self.make_token(TokenKind::RightBracket),
            '{' => self.make_token(TokenKind::LeftBrace),
            '}' => self.make_token(TokenKind::RightBrace),
            ';' => self.make_token(TokenKind::Semicolon),
            ',' => self.make_token(TokenKind::Comma),
            '.' => self.make_token(TokenKind::Dot),
            '-' => {
                let token = if self.check_next('>') {
                    TokenKind::Arrow
                } else if self.check_next('=') {
                    TokenKind::MinusEqual
                } else {
                    TokenKind::Minus
                };
                self.make_token(token)
            }
            '+' => {
                let token = if self.check_next('=') {
                    TokenKind::PlusEqual
                } else {
                    TokenKind::Plus
                };
                self.make_token(token)
            }
            '/' => {
                let token = if self.check_next('/') {
                    while self.peek().is_some() && self.peek().unwrap() != '\n' && !self.is_at_end()
                    {
                        self.advance();
                    }
                    TokenKind::Comment
                } else {
                    TokenKind::Slash
                };
                self.make_token(token)
            }
            '*' => self.make_token(TokenKind::Star),
            ':' => self.make_token(TokenKind::Colon),
            '%' => self.make_token(TokenKind::Percent),
            '!' => {
                let token = if self.check_next('=') {
                    TokenKind::BangEqual
                } else {
                    TokenKind::Bang
                };
                self.make_token(token)
            }
            '=' => {
                let token = if self.check_next('=') {
                    TokenKind::EqualEqual
                } else {
                    TokenKind::Equal
                };
                self.make_token(token)
            }
            '<' => {
                let token = if self.check_next('=') {
                    TokenKind::LessEqual
                } else {
                    TokenKind::Less
                };
                self.make_token(token)
            }
            '>' => {
                let token = if self.check_next('=') {
                    TokenKind::GreaterEqual
                } else {
                    TokenKind::Greater
                };
                self.make_token(token)
            }
            '"' => self.string(),
            _ => Err(self.err("Unexpected character".to_string())),
        }
    }

    fn identifier(&mut self) -> Result<Token> {
        while self
            .peek()
            .is_some_and(|c| c.is_ascii_alphabetic() || c.is_digit(10) || c == '_')
        {
            self.advance();
        }
        self.make_token(self.identifier_kind())
    }

    fn identifier_kind(&self) -> TokenKind {
        match TokenKind::from_str(&self.source[self.start..self.current]) {
            Ok(kind) => kind,
            Err(_) => TokenKind::Identifier(self.source[self.start..self.current].to_string()),
        }
    }

    fn number(&mut self) -> Result<Token> {
        while self.peek().is_some()
            && (self.peek().unwrap().is_digit(10) || self.peek().unwrap().eq(&'.'))
        {
            self.advance();
        }
        self.make_token(TokenKind::Number(
            self.source
                .get(self.start..self.current)
                .unwrap()
                .parse()
                .unwrap(),
        ))
    }

    fn string(&mut self) -> Result<Token> {
        let line = self.line;
        while self.peek().is_some() && self.peek().unwrap() != '"' && !self.is_at_end() {
            if self.peek().unwrap() == '\n' {
                self.line += 1;
                self.advance();
            }
            self.advance();
        }
        if self.is_at_end() {
            return Err(self.err("Unterminated string".to_string()));
            // return Err(Error {
            //     line,
            //     kind: ErrorKind::UnterminatedString,
            //     cols: Some((self.start, self.current)),
            // });
        }
        self.advance();
        self.make_token(TokenKind::String(
            self.source
                .get((self.start + 1)..(self.current - 1))
                .unwrap()
                .to_string(),
        ))
    }

    fn skip_whitespace(&mut self) -> Result<()> {
        loop {
            if let Some(c) = self.peek() {
                match c {
                    ' ' | '\r' | '\t' => _ = self.advance(),
                    '\n' => {
                        self.line += 1;
                        self.advance();
                        self.column = 0;
                    }
                    '#' => {
                        self.advance();
                        while self.peek().is_some() && self.peek().unwrap() != '#' {
                            self.advance();
                        }
                        self.advance();
                    }
                    _ => break,
                }
            } else {
                break;
            }
        }
        Ok(())
    }

    fn peek_next(&self) -> Option<char> {
        if self.is_at_end() {
            return None;
        }
        self.source.chars().nth(self.current + 1)
    }

    fn peek(&self) -> Option<char> {
        self.source.chars().nth(self.current)
    }

    fn check_next(&mut self, c: char) -> bool {
        if self.is_at_end() {
            return false;
        }
        if self.source.chars().nth(self.current).unwrap() != c {
            return false;
        }
        self.current += 1;
        self.column += 1;
        true
    }

    fn advance(&mut self) -> Option<char> {
        self.column += 1;
        self.current += 1;
        self.source.chars().nth(self.current - 1)
    }

    fn make_token(&self, kind: TokenKind) -> Result<Token> {
        Ok(Token {
            kind,
            line: self.line,
            end: self.column,
            start: self.column - (self.current - self.start),
        })
    }

    fn is_at_end(&self) -> bool {
        self.current >= self.source.len()
    }
}

File: src/Scanner.cs
class Scanner
{
    private int _start;
    private int _current;
    private int _line;
    private int _col;
    private int _startCol;

    private Dictionary<string, TokenKind> _twoCharTokens;

    private string _source;


    private char CurrentChar()
    {
        return _source[_current];
    }

    public string Source => _source;

    public Scanner(string source)
    {
        _start = _current = _line = _col = _startCol = 0;
        _source = source.Replace("\r", "");

        _twoCharTokens = new()
        {
            ["!="] = TokenKind.BangEqual,
            ["=="] = TokenKind.EqualEqual,
            [">="] = TokenKind.GreaterEqual,
            ["<="] = TokenKind.LessEqual,
            ["->"] = TokenKind.Arrow,
            ["+="] = TokenKind.PlusEqual,
            ["-="] = TokenKind.MinusEqual,
        };
    }

    /// <summary>
    /// Parses all the text in the Scanner to tokens.
    /// </summary>
    /// <returns>
    /// A IEnumerable of the tokens representing the file content. The last token is
    /// always an 'EndOfFile' token. 
    /// </returns>
    public IEnumerable<Token> Tokens()
    {
        while (!IsAtEnd())
            yield return NextToken();
    }

    private Token MakeToken(TokenKind kind) =>
        new Token(
            kind,
            _source[_start.._current],
            _line,
            _startCol,
            _startCol + (_current - _start)
        );

    private Token NextToken()
    {
        SkipWhitespace();
        _start = _current;
        _startCol = _col;
        if (IsAtEnd())
            return MakeToken(TokenKind.EndOfFile);
        var c = Advance();
        return c switch
        {
            >= 'a' and <= 'z' or >= 'A' and <= 'Z' => Identifier(),
            >= '0' and <= '9' => Number(),
            '(' => MakeToken(TokenKind.LeftParen),
            ')' => MakeToken(TokenKind.RightParen),
            '[' => MakeToken(TokenKind.LeftBracket),
            ']' => MakeToken(TokenKind.RightBracket),
            '{' => MakeToken(TokenKind.LeftBrace),
            '}' => MakeToken(TokenKind.RightBrace),
            ';' => MakeToken(TokenKind.Semicolon),
            ',' => MakeToken(TokenKind.Comma),
            '.' => MakeToken(TokenKind.Dot),
            '/' => MakeToken(Comment()),
            '*' => MakeToken(TokenKind.Star),
            ':' => MakeToken(TokenKind.Colon),
            '%' => MakeToken(TokenKind.Percent),
            '|' => MakeToken(TokenKind.Pipe),
            '-' => MakeToken(TwoCharTokens(c, TokenKind.Minus)),
            '+' => MakeToken(TwoCharTokens(c, TokenKind.Plus)),
            '!' => MakeToken(TwoCharTokens(c, TokenKind.Bang)),
            '=' => MakeToken(TwoCharTokens(c, TokenKind.Equal)),
            '<' => MakeToken(TwoCharTokens(c, TokenKind.Less)),
            '>' => MakeToken(TwoCharTokens(c, TokenKind.Greater)),
            '"' => String(),
            _ => throw new SparvException($"Unexpected char '{c}'", _line, _startCol, _current),
        };
    }

    private TokenKind Comment()
    {
        if (CurrentChar() != '/')
            return TokenKind.Slash;
        while (!IsAtEnd() && CurrentChar() != '\n') Advance();
        return TokenKind.Comment;
    }

    private TokenKind TwoCharTokens(char first, TokenKind single)
    {
        var second = CurrentChar();
        var s = $"{first}{second}";
        if (_twoCharTokens.ContainsKey(s))
        {
            Advance();
            return _twoCharTokens[s];
        }
        return single;
    }

    private TokenKind Keywords(string s)
    {
        return s switch
        {
            "var" => TokenKind.Var,
            "true" => TokenKind.True,
            "and" => TokenKind.And,
            "else" => TokenKind.Else,
            "if" => TokenKind.If,
            "nil" => TokenKind.Nil,
            "or" => TokenKind.Or,
            "return" => TokenKind.Return,
            "while" => TokenKind.While,
            "false" => TokenKind.False,
            "for" => TokenKind.For,
            "loop" => TokenKind.Loop,
            "fun" => TokenKind.Fun,
            "in" => TokenKind.In,
            "import" => TokenKind.Import,
            "match" => TokenKind.Match,
            _ => TokenKind.Identifier
        };
    }

    private Token Identifier()
    {
        while (!IsAtEnd())
        {
            var c = CurrentChar();
            if (!char.IsAsciiLetter(c) && !char.IsNumber(c) && c != '_')
            {
                break;
            }
            Advance();
        }
        var kind = Keywords(_source[_start.._current]);
        return MakeToken(kind);
    }
    private Token Number()
    {
        while (!IsAtEnd())
        {
            var c = CurrentChar();
            if (!char.IsNumber(c) && c != '_' && c != '.')
            {
                break;
            }
            Advance();
        }
        return MakeToken(TokenKind.Number);
    }
    private Token String()
    {
        while (!IsAtEnd() && Advance() != '"') ;
        if (IsAtEnd()) throw new SparvException("Unterminated string", _line, _startCol, _col);
        return MakeToken(TokenKind.String);
    }

    private bool IsAtEnd() => _current >= _source.Length;

    private void SkipWhitespace()
    {
        while (!IsAtEnd())
        {
            var c = CurrentChar();
            if (c == '\n')
            {
                _line += 1;
                _col = 0;
                _current += 1;
            }
            else if (char.IsWhiteSpace(c))
            {
                Advance();
            }
            else
            {
                break;
            }
        }
    }

    private char Advance()
    {
        var c = CurrentChar();
        _current += 1;
        _col += 1;
        return c;
    }

}

public record Token(TokenKind Kind, string Value, int Line, int Start, int End);

public enum TokenKind
{
    // Single-character tokens.
    /// <summary> TokenKind: ( </summary>
    LeftParen,
    /// <summary> TokenKind: ) </summary>
    RightParen,
    /// <summary> TokenKind: [ </summary>
    LeftBracket,
    /// <summary> TokenKind: ] </summary>
    RightBracket,
    /// <summary> TokenKind: { </summary>
    LeftBrace,
    /// <summary> TokenKind: } </summary>
    RightBrace,
    Comma,
    Dot,
    Minus,
    Plus,
    Semicolon,
    Slash,
    Star,
    Colon,
    Percent,
    Pipe,
    // One or two character tokens.
    Bang,
    BangEqual,
    Equal,
    EqualEqual,
    Greater,
    GreaterEqual,
    Less,
    LessEqual,
    Arrow,
    PlusEqual,
    MinusEqual,
    // Literals.
    Identifier,
    String,
    Number,
    // Keywords.
    Var,
    And,
    Else,
    False,
    For,
    Loop,
    Fun,
    If,
    Nil,
    Or,
    Return,
    True,
    In,
    While,
    Import,
    Match,
    EndOfFile,
    // Comments
    Comment,
}
